### ISSUE TITLE
Failed to construct ACL on failover

### ISSUE THREAD

**Issue by wneessen** [hashicorp/vault#7214]
**Describe the bug**
I am running a Vault-HA setup (3 consul nodes as storage, 3 vault servers on each node). When I set up a fresh vault installation, at the beginning all looks fine. Each node is in a correct state, consul is working fine and a "consul monitor" does not throw any errors. I can unseal the cluster and also log in with the root token. When I create a user and assign that user it's policies, I will be able to log in with that user and do whatever it is supposed to do. All fine. But once I perform a "vault operate step-down" on the active instance and I try to log in with the same user again, I run into problems. The login still works, but each access gives a HTTP 500. In the vault server logs I receive the following message:

Jul 30 13:52:49 vault3.xxxxxx vault[39356]: 2019-07-30T13:52:49.134Z [ERROR] core: failed to construct ACL: error="failed to get policy: failed to read policy: decryption failed: cipher: message authentication failed"
Jul 30 13:52:49 vault3.xxxxxx vault[39356]: 2019-07-30T13:52:49.170Z [ERROR] core: failed to construct ACL: error="failed to get policy: failed to read policy: decryption failed: cipher: message authentication failed"
Jul 30 13:52:51 vault3.xxxxxx vault[39356]: 2019-07-30T13:52:51.496Z [ERROR] core: failed to construct ACL: error="failed to get policy: failed to read policy: decryption failed: cipher: message authentication failed"

I was running the same installation for a couple of month without any issues, I then did OS updates (and program package updates - no consul or vault package update was included) on all nodes and it seem that since then the issue occured. I was under the impression that my consul wasn't working fine, so I reset the whole consul setup and started from scratch with a fresh consul and vault installation, but the same occurred again after the first "fail-over" from an active to a standby instance.

I can in fact use the root token to log in and everything work, I can even overwrite the ACLs and the user will work again, but as soon as I fail over again, it will stop working. I assume this is some kind of bug as this happens even with a freshly set up new instance.

**To Reproduce**
1. Install a 3 node vault/consul cluster (as described here: https://learn.hashicorp.com/vault/operations/ops-vault-ha-consul )
2. Create a fresh vault instance
3. Unseal all 3 instances
4. Create auth -> userpass
5. Create ACLs
6. Create a userpass-user with the new ACL assigned
7. Log into Vault with that user -> all should be fine
8. Issue "vault operate step-down"
9. Log into the new active Vault node -> Receive HTTP 500s

**Expected behavior**
The failover should work as expected without corrupting the ACLs

**Environment:**
* Vault Server Version (retrieve with `vault status`): 1.1.2
* Vault CLI Version (retrieve with `vault version`): 1.1.2
* Server Operating System/Architecture: 3 x FreeBSD 12.0-RELEASE-p8
* Consul version: 1.5.1

Vault server configuration file(s):

```hcl
listener "tcp" {
  address          = "0.0.0.0:8200"
  cluster_address  = "vault1.xxxxxx:8201"
  tls_cert_file = "/usr/local/etc/vault.d/vault1.xxxxxx.crt"
  tls_key_file = "/usr/local/etc/vault.d/vault1.xxxxxx.key"
  tls_ca_file = "/usr/local/etc/vault.d/corpCA_ServerCA.cabundle"
  tls_disable_client_certs = 1
}

storage "consul" {
  address = "127.0.0.1:8501"
  path    = "vault/"
  scheme  = "https"
  tls_ca_file = "/usr/local/etc/vault.d/corpCA_ServerCA.cabundle"
}

api_addr = "https://vault1.xxxxxx:8200"
cluster_address  = "https://vault1.xxxxxx:8201"
disable_mlock = 1
ui = 1
```

**Comment by wneessen** [comment]
Meanwhile I updated to the latest releases of Vault and Consul (1.2.2 and 1.6.0) and also upgraded the OS on all nodes to 12-RELEASE-p10. Still the issue occurs. Any help to get this bug fixed?

**Comment by michelvocks** [comment]
Hi @wneessen,

I tried to reproduce your issue, but I wasn't successful.
In my opinion, only two things could be responsible for such a problem to appear: 

- Consul is in a bad state. If one instance steppes down, it actually removes a lock (e.g. an entry in Consul) and the first instance which is able to grep the lock again (e.g. create that entry) is the leader. I assume that something went wrong during that process. Did you check the Consul logs? Any suspicious log lines are existing?
- Vault is somehow wrongly configured. I noticed that you have defined `cluster_address` in your Vault file which should be `cluster_addr` (https://www.vaultproject.io/docs/configuration/#cluster_addr).

Cheers,
Michel

**Comment by wneessen** [comment]
Hi @michelvocks,

Nothing in the consul logs. I even set up a fresh consul cluster and a fresh vault instance and only setup the "superuser" policy. Same behaviour. I can log in with the root token, I can log in with my "superuser" but once I perform a "operator step-down", the superuser has no access anymore, the root user still works. Once I log in with the root token again, click on "edit" of the superuser policy, do no change but just save it again, the superuser will be able to log in again.

I also corrected the cluster_addr setting in the config file (thanks for the hint). Still the same result.

**Comment by michelvocks** [comment]
Could you provide the Vault configuration files from the other instances? Can you verify that the ports are not blocked by a firewall between the hosts? In general, more information about your setup would be helpful e.g. Consul setup (where are the servers, how many etc.), infrastructure setup e.g. connection between hosts. 

Cheers,
Michel

**Comment by wneessen** [comment]
Absolutely! Is there a secure way I can get those files to you? I'd rather not provide everything in this public  ticket.

**Comment by michelvocks** [comment]
You can DM me or send them to me via email: michel@hashicorp.com

**Comment by michelvocks** [comment]
Hi @wneessen!

I tried to reproduce your issue with a fresh FreeBSD 12.0 setup including the configuration you have send to me via E-Mail. Sadly, I still wasn't able to reproduce your mentioned behavior.

I'm a bit clueless on how to proceed here. I assume it has something to do with your network setup or with your used certificates which is hard to reproduce on our side. I don't think this is a FreeBSD/Vault bug otherwise I would be able to reproduce it.

Is there any other information available e.g. certificates info (maybe a cert has been expired?) or monitoring information about the network?

Cheers,
Michel

**Comment by michelvocks** [comment]
Hi @wneessen!

I'm closing this due to inactivity. Feel free to open a new issue if this is still a thing for you!

Cheers,
Michel

**Comment by wneessen** [comment]
The issue has been identified via https://github.com/hashicorp/vault/issues/8760. Turned out it was related the the policy naming and not related to the syncing of the storage.
